fail-safely.md

```diff
@@ -0,0 +1,100 @@
+
+```go
+if user.State == Deleted {
+  return someErr
+}
+```
+
+which is doomed to fail open. What happens when a new state is added? Yep, it's _implicitly_ allowed. Congratulations, the new `Purged` state is now allowed where `Deleted` was not.
+
+We can make it a little better by describing what we want to continue, and negating it. See also [Happy Path Expressions](./negated-happy-path-form.md).
+
+```go
+if user.State != Active {
+  return someErr
+}
+```
+
+While this is _better_, we're still _implicitly_ denying any other type, but we're talking about a known set of values here. The best thing to do would be to exhaustively handle every related state, _explicitly_ allowing or denying each one, and by doing so, protecting against that future new state with a default deny.
+
+```go
+switch user.State {
+  case Active, Pending:
+    // okay, continue...
+
+  case Deleted:
+    return someErr
+
+  default:
+    return errNotHandled
+}
+```
+
+Go doesn't have enums, so I don't know of a way to easily check for this at compile time, but the next best thing is at runtime. At least something like this will help prevent a user getting into a weird state. Debugging that is way harder than debugging the `errNotHandled` default deny.
+
+An additional benefit is that the code will very readily point out all places that need to be considered when adding a new state, either by finding all of the locations via an editor, or at runtime when it throws errors.
+
+Do this for all related would-be enums: states, types, kinds, etc.
+
+There's one more example, and that is the explicit fail-open form, and it's not as uncommon as you might think. It's particularly useful in situations where it's still safe to proceed even if the value is unrecognized. This can be a good strategy when introducing something new and is riskier to fail closed unnecessarily than allow execution to continue.
+
+```go
+switch user.State {
+  case Active, Pending:
+    // okay, continue...
+
+  case Deleted:
+    return someErr
+
+  default:
+    log.Println("unknown user state %q, continuing")
+}
+```
+
+That example is a little more difficult to see it's usefulness but I wanted to include it for the ability to compare the examples that all use the same set of values. Here's a more realistic example of explicitly failing open. Say I have a process running and it's listening for signals. If we recieve an unrecognized signal, it should just be ignored.
+
+```go
+switch cmd {
+case Start:
+  // start the process
+
+case Stop:
+  // stop the process
+
+case Restart:
+  // restart the process
+
+default:
+  fmt.Printf("Unknown command: %s\n", cmd)
+  // continue...
+}
+```
+
+And that's really it. Avoid implictly failing open, and beyond that choose the right tool for the job. When I'm dealing with a set of related values, I typically choose a switch block. It results in more lines of code, but I believe the trade off for explicitness and debuggability are worth it.
```

Comment:
I’m really starting to wonder when implicitness is ever a good thing. I guess only when a piece of information is so commonly present in some specific context would explicitness add more noise than value. But otherwise, and especially in type definitions, the more explicit you can be, the safer your code.

Reply:
I'd be willing to argue that the only form to always avoid is the implicit fail open.

> I guess only when a piece of information is so commonly present in some specific context would explicitness add more noise than value.

If by more noise you mean more text on screen, I disagree that it's noise. At a certain subjective point, a conditional is simpler to understand and change when it goes from an `if` block to a `switch` block. It's fine to start with the implicit fail closed form and upgrade to explicit.

One benefit of being explicit is that it's more likely that someone thought about it rather than a certain case being missed and it just happens to work.

But yes, it's common for these ideas to get feedback that they feel like more code than necessary. As long as folks are deliberately choosing one form over the other, I think that's a win.

Reply:
> If by more noise you mean more text on screen, I disagree that it's noise.

I wasn't implying this was a scenario that created noise, I was considering implicit/explicit information about a program _in general_.

Memory allocation for example, since go is garbage collected, is fairly implicit in go programs, and that seems like a good thing. Though that has nothing to do with your business logic, so maybe that's the line.

Just thinking! We're aligned here. The `switch` encourages more a more explicit expression of the logic, which is a positive.